### PR TITLE
feat(portal): Add Microsoft Entra IDP sync to portal

### PR DESCRIPTION
**Pull request by bmanifold**
Why:

* To allow syncing of users/groups/memberships from an IDP to Firezone, a custom identify provider adapter needs to be created in the portal codebase at this time.  The custom IDP adapter created in this commit is for Microsoft Entra.

### CODE REVIEW

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex:1]: Kudos to @AndrewDryga for making the sync system modular and relatively straightforward to extend to other providers

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+          "$select" =>
+            Enum.join(
+              ~w[
+                id
+                accountEnabled
+                displayName
+                givenName
+                surname
+                mail
+                userPrincipalName
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex]: Would it be worth making this a module attribute `@keys_to_fetch`?

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+              ],
+              ","
+            ),
+          "$filter" => "accountEnabled eq true"
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex:61]: Same here re: module attribute.

Companies have wanted to apply user email filtering in the past, so this could even be pulled from config set via env var.

This is probably fine for now, we can refactor if the need arises.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+          "$select" =>
+            Enum.join(
+              ~w[
+                id
+                accountEnabled
+                description
+                displayName
+              ],
+              ","
+            )
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex]: Would it be worth adding the `accountEnabled` filter here too?

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex [hunk 4]
```diff
@@ -0,0 +1,201 @@
+defmodule Domain.Auth.Adapters.MicrosoftEntra.Jobs do
+  use Domain.Jobs.Recurrent, otp_app: :domain
+  alias Domain.{Auth, Actors}
+  alias Domain.Auth.Adapters.MicrosoftEntra
+  require Logger
+
+  every minutes(5), :refresh_access_tokens do
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:8]: Would it be worth pulling this from config/env so we can just change it in terraform without having to redeploy?

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+              provider_id: provider.id,
+              account_id: provider.account_id,
+              reason: inspect(reason)
+            )
+        end
+      end)
+    end
+  end
+
+  every minutes(3), :sync_directory do
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:37]: Same here re: making this runtime-configurable

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+      }
+    end)
+  end
+
+  # Map group attributes from Microsoft Entra to Domain
+  defp map_group_attrs(groups) do
+    Enum.map(groups, fn group ->
+      %{
+        "name" => "Group:" <> group["displayName"],
+        "provider_identifier" => "G:" <> group["id"]
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:166]: Is the `G` for Google?

```suggestion
        "provider_identifier" => "M:" <> group["id"]
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/microsoft_entra/components.ex [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    </div>
+    """
+  end
+
+  def scopes do
+    """
+    openid
+    profile
+    email
+    offline_access
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/microsoft_entra/components.ex:116]: I was curious about this one -- does requesting this give us the long-lived refresh token we need to perform this in the background?

I seem to remember Google behaving a bit differently

**Review by jamilbk — COMMENTED** [review]
Looking good so far

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex (hunk 1 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex]: Yeah, I can do that.  Make one for users and one for groups, since we aren't pulling the same attributes for both.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex (hunk 2 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex:61]: I think it would be good to have some flexibility in what gets pulled/synced from the different IDPs and we probably do want to make this some type of variable rather than a hard coded string.  I'll move this to a module attribute.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex (hunk 3 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex]: Possibly.  I thought I had tested users being removed from a group when they were set to `inactive`, but I need to double check.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 4 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:8]: I think it would be a good idea to be able to control this timing through some type of config.  That said, if we put something in terraform, isn't that going to force a redeploy due to how our deployment works?

Ideally I could see this being set by some type of Admin Ops dashboard in the future, but that may be a ways off.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 6 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:166]: I was thinking the `G` stood for Group, but let me double check that.

File: elixir/apps/web/lib/web/live/settings/identity_providers/microsoft_entra/components.ex (hunk 7 above)
Comment (bmanifold) [reply elixir/apps/web/lib/web/live/settings/identity_providers/microsoft_entra/components.ex:116]: Yeah, if we don't have the `offline_access` scope then we don't get a refresh token and we'd only be able to sync as long as the original access token was valid (which is not all that long)

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 6 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:166]: `G` is for `Group`. Because google has "OU:". But if Entra doesn't have any other entities we can remove the prefix.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 5 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:37]: Let's not do this for now, especially in the scope of this PR.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 4 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:8]: Brian is right, redeploying and changing one line with a typical workflow is the same "time to deliver", only difference is that somebody can go manually do `terraform apply` to update it in case of emergencies, which, I think, not worth it now as there are other ways to do it with Elixir.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex (hunk 2 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex:61]: I think in future we should just add some UI for the provider settings and allow them to extend the filters, but `accountEnabled` should be always true, there is no point to sync trash data for us - they will never be able to sign in.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 4 above)
Comment (jamilbk) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:8]: Yeah forgot about the apply step, was thinking it was just an env read on boot

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 6 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex:166]: I decided to keep the `G:` prefix for now.  There are organizational units in Entra, but I'm not quite clear on whether a single tenant would actually use them.  If it turns out they do then having the `G:` prefix will be good and if they don't, then the `G:` prefix won't get in the way.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex (hunk 3 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/api_client.ex]: Since this is outdated, it's not show me the whole code snippet, but if this was for the group membership call, I put a comment in the code as to why I didn't include the `$filter` on that call.

File: .tool-versions [hunk 8]
```diff
@@ -3,7 +3,7 @@
 nodejs 18.19.0
 elixir 1.15.7-otp-26
 erlang 26.1.2
-terraform 1.6.6
+#terraform 1.6.6
```
Comment (AndrewDryga) [inline .tool-versions]: Is this change intentional?

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      {Domain.Jobs, MicrosoftEntra.Jobs}
+    ]
+
+    Supervisor.init(children, strategy: :one_for_one)
+  end
+
+  @impl true
+  def capabilities do
+    [
+      provisioners: [:just_in_time, :custom],
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex]: Please remove JIT from this list, I'll do that in other places in a different PR.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+      }
+    end)
+  end
+
+  # Log effects of the multi transaction
+  defp log_effects(provider, effects) do
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex]: Should we move it to some shared place?

**Review by AndrewDryga — APPROVED** [review]
A few little things and LGTM

File: .tool-versions (hunk 8 above)
Comment (bmanifold) [reply .tool-versions]: Ah, good catch.  Nope, was having some issues with `asdf` and was just removing that from the equation.  Will get this put back to what it was.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex (hunk 9 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex]: 👍 sounds good

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex (hunk 10 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex]: Yeah, I thought about that just today, but then forgot during some of my test debugging.  I'll try to get this DRY'd up.

Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex]: Wasn't exactly sure where to put this code, so just took a stab at it.  Let me know if this isn't a good convention or just doesn't fit in some way.

File: .tool-versions (hunk 8 above)
Comment (bmanifold) [reply .tool-versions]: Got this changed back to what's in main.

File: elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex (hunk 9 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra.ex]: Removed the `:just_in_time` option.

**Review by jamilbk — APPROVED** [review]
LGTM, excited to test this out!
